### DISCUSSION TITLE
Basic understanding: blocking access to URIs from specific groups

### DISCUSSION THREAD

**Discussion by rcapp** [keycloak/keycloak#16204] [Q&A] [answered]
Hello, everyone!

I'm quite new to Keycloak and I still have some beginner doubts and questions. Specially regarding Keycloak's capabilities and how it works as an Authorizer. 

1. Can we block access to *client's* specific URIs? 

We have one API, that's configured under a client inside our Realm. For better understanding, I will call it  `User API`. In this `User API`  I can access their information, like home address, telephone, last purchases, average expanses, etc.  This information I can get at `/api/user-info`.   

Suppose only the user and people at `platform_manager` or `platform_admin` can access this URL. In my understanding, I would create a group/role and add users there. Then, I would create go to `Clients > Client details >  Create resource` and fill the fields `Authorization scopes` and `URI` (`/api/user-info`) with the wanted data. 

Even with these filled up, I can access User API without authentication. I know the authentication works because I can get  my tokens, do a browser login and validate my token via my API's backend. 

My goal would be to block any access *before* the backend is reached at that route, like when we try to authenticate with an invalid token:
```
{
    "detail": "Token verification failed"
}
```

2. When a field in Keycloak asks for an URI, is this related with my API/Application or is this Keycloak's own address? (i.e. you block access to an API address from Keycloak, not yours)

3. If that is possible, how can we do that? I don't intend to use something like `OAuth Proxy` because if I understood it right, it would work similar to a proxy running before my backend (and we already have one). Is it right? 

4. I tried to use `Auth type` with `confidential` and it did not change a thing on this process. What does this actually do?

Bonus: Where the `Authentication type` sits on ver. 20? I had to change the theme to find this option. 
 

Thank you in advance!

**Comment by sashokbg** [comment, marked as answer]
Hello, you should check the keycloak quistarters repo.

Example: https://github.com/keycloak/keycloak-quickstarts/tree/latest/app-authz-spring-security

There they show you exactly how to protect a resource (URL) based on couple of different criteria - either policy, owner attribute or role.

I also recently worked on a POC of this and I have created a terraform script that prepared keycloak, so if you want you can check my repo: https://github.com/sashokbg/keycloak-authorization

**Reply by rcapp** [reply]
Thanks for point out this example! 

Checking it I was able to follow that file and build something similar. However, I can't find  the mentioned file `script-always-grant.js`. I'm also not able to create a `JS` Policy. What am I missing?

Cheers!

**Reply by sashokbg** [reply]
JS policies are now (since some recent version) deployed as a standalone JAR.

Check the https://github.com/keycloak/keycloak-quickstarts/tree/latest/authz-js-policies

Basicaly you build the "authz-js-policies" and copy it to the keycloak install dir:  keycloak/providers/js-policies.jar

Or if you follow my repo it is mounted using docker volume

**Reply by rcapp** [reply]
Hi! Sorry for the late reply.

I'm still trying to figure out how to use it in our scenario, but so far we decided to validate the paths accessed via our API, based on what the `evaluate` function outputs (via UMA).  So far, it is sufficing ours needs. We intend to do more complex (and probably safer) methods of authorization in the near future. 

Thanks for the help!
